src/Features/Core/Portable/AddConstructorParametersFromMembers/AddConstructorParametersFromMembersCodeRefactoringProvider.cs

```diff
@@ -63,79 +72,75 @@ public static async Task<ImmutableArray<CodeAction>> AddConstructorParametersFro
-                        containingType,
-                        constructorCandidate.MissingParameters,
-                        useSubMenuName: false));
+                    actions.Add(new CodeAction.CodeActionWithNestedActions(
+                        FeaturesResources.Add_parameter_to_constructor,
+                        result.RequiredParameterActions.Cast<AddConstructorParametersCodeAction, CodeAction>(),
+                        isInlinable: false));
                 }
 
-                result.Add(GetOptionalContructorParametersCodeAction(
-                    document,
-                    constructorCandidate,
-                    containingType,
-                    useSubMenuName: false));
+                actions.Add(new CodeAction.CodeActionWithNestedActions(
+                    FeaturesResources.Add_optional_parameter_to_constructor,
+                    result.OptionalParameterActions.Cast<AddConstructorParametersCodeAction, CodeAction>(),
+                    isInlinable: false));
             }
             else
             {
-                // Create sub menus for suggested actions, one for required parameters and one for optional parameters
-                using var _1 = ArrayBuilder<CodeAction>.GetInstance(out var requiredParameterCodeActions);
-                using var _2 = ArrayBuilder<CodeAction>.GetInstance(out var optionalParameterCodeActions);
-                foreach (var constructorCandidate in state.ConstructorCandidates)
+                // Not using submenus, this means we have at most a single of each action.
+                if (!result.RequiredParameterActions.IsDefaultOrEmpty)
                 {
-                    if (CanHaveRequiredParameters(constructorCandidate.Constructor.Parameters))
-                    {
-                        requiredParameterCodeActions.Add(new AddConstructorParametersCodeAction(
-                            document,
-                            constructorCandidate,
-                            containingType,
-                            constructorCandidate.MissingParameters,
-                            useSubMenuName: true));
-                    }
+                    actions.Add(result.RequiredParameterActions.Single());
+                }
+
+                actions.Add(result.OptionalParameterActions.Single());
+            }
 
-                    optionalParameterCodeActions.Add(GetOptionalContructorParametersCodeAction(
+            return actions.ToImmutable();
+        }
+
+        private static AddConstructorParameterResult CreateCodeActions(Document document, State state)
+        {
+            using var _0 = ArrayBuilder<AddConstructorParametersCodeAction>.GetInstance(out var requiredParametersActions);
+            using var _1 = ArrayBuilder<AddConstructorParametersCodeAction>.GetInstance(out var optionalParametersActions);
+            var containingType = state.ContainingType;
+
+            var useSubMenu = state.ConstructorCandidates.Length > 1;
```

Comment:
is this value necessary?  would it be implied by having multiple actions versus a single action?  (i haven't looked closely.  it may be necessary due to it meaning more than that.  but if we can remove and just use the aciton count, that would be nice).

Reply:
I preferred this way because the action counts are different between the requiredparameteractions and optionalparameteractions.  I could use OptionalParameterActions as that one does match ctor candidates length, but it seemed like it could be confusing as to why I only check the OptionalParameterActions length.  I'm open switching though

Reply:
totally up to you.  :)